doc/user/interop.md

```diff
@@ -89,150 +90,113 @@ Returns true only for the `nil` object.
+- `INSERTABLE` and `MODIFIABLE` will be set if the object has a `[]=` method.
 
-For all objects and keys:
+- `REMOVABLE` and `INTERNAL` will not be set.
 
-`INVOCABLE` is never set, because currently `KEYS` does not include methods.
+- `INVOCABLE` will be set if the object has a method of that name.
 
-`EXISTING` is set if `READABLE`, `MODIFIABLE`, `INVOCABLE`, `INTERNAL` or
-`REMOVABLE` are set.
+For all objects:
 
-`REMOVABLE` is always set to false.
+- `EXISTING` is set if `READABLE`, `MODIFIABLE`, `INVOCABLE`, `INTERNAL` or
+  `REMOVABLE` are set.
 
 ### `READ`
 
-The name must be an `int` (small Ruby `Fixnum`), or a Ruby `String` or `Symbol`,
-or a Java `String`.
-
-If the receiver is a Ruby `String` and the name is an integer, read a byte from
-the string, ignoring the encoding. If the index is out of range you'll get 0:
-
-`READ(string, integer)` → `string.get_byte(integer)`
+If the receiver is a Ruby `String`, return bytes in the string as if
+`receiver.bytes[n]` was used, or return `0` if the byte is out of range.
 
-Otherwise, if the name starts with `@`, read it as an instance variable:
+If the receiver is a Ruby `Array` or `Hash`, call `receiver[name]`.
 
-`READ(object, "@name")` → `object.instance_variable_get("name")`
+Otherwise if the name starts with an `@` it is read as an instance variable.
 
-Otherwise, if there is a method defined on the object with the same name as
-the name, perform a method call using the name as the called method name:
+Otherwise, if there is a method called `[]` on the receiver, call
+`receiver[name]`.
 
-`READ(object, name)` → `object.name` if `object.responds_to?(name)`
+Otherwise, if there is a method defined on the object with the same name, return
+it as a (bound) `Method`.
 
-Otherwise, if there isn't a method defined on the object with the same name as
-the name, and there is a method defined on the object called `[]`, call `[]`
-with the name as the argument:
-
-`READ(object, name)` → `object[name]` unless `object.responds_to?(name)`
-
-Otherwise throws `UnknownIdentifierException`.
-
-In all cases where a call is made no block is passed.
-
-An exception during a read will result in an `UnknownIdentifierException`.
+Otherwise, throw `UnknownIdentifierException`.
 
 ### `WRITE`
 
-The name must be a `String` or `Symbol`.
-
-If the name starts with `@`, write it as an instance variable:
-
-`WRITE(object, "@name", value)` → `object.instance_variable_set("name", value)`
+If the receiver is a Ruby `Array` or `Hash`, call `receiver[name] = value`.
 
-Otherwise, if there is a method defined on the object with the same name as
-the name appended with `=`, perform a method call using the name appended with
-`=` as the called method name, and the value as the argument:
+Otherwise if the name starts with an `@` it is set as an instance variable.
 
-`WRITE(object, name, value)` → `object.name = value` if
-`object.responds_to?(name + "=")`
+Otherwise, if there is a method called `[]=` on the receiver, call
```

Comment:
I think this is problematic for `String`. The `READ` message special-cases `String` basically call `String#getbyte`. There's no corresponding special case for `WRITE`. However, `String` does have a `[]=` method and that method operates at the character level. I think we should also special-case `WRITE` for `String` to avoid potentially confusing situations and just make the entire thing byte-oriented.

Alternatively, we should remove the special-case for `READ` of `String`s and allow `String#[]` to be called. This will be character-oriented, but on binary strings that boils down to being byte-oriented anyway.

Reply:
I've removed the special logic for `READ` of a `String`.